modules/calib3d/src/usac/utils.cpp

```diff
@@ -0,0 +1,229 @@
+    const auto * const points_ = (float *) points.data;
+    const auto * const k1 = (double *) K1.data;
+    const auto inv1_k11 = float(1 / k1[0]); // 1 / k11
+    const auto inv1_k12 = float(-k1[1] / (k1[0]*k1[4])); // -k12 / (k11*k22)
+    // (-k13*k22 + k12*k23) / (k11*k22)
+    const auto inv1_k13 = float((-k1[2]*k1[4] + k1[1]*k1[5]) / (k1[0]*k1[4]));
+    const auto inv1_k22 = float(1 / k1[4]); // 1 / k22
+    const auto inv1_k23 = float(-k1[5] / k1[4]); // -k23 / k22
+
+    const auto * const k2 = (double *) K2.data;
+    const auto inv2_k11 = float(1 / k2[0]);
+    const auto inv2_k12 = float(-k2[1] / (k2[0]*k2[4]));
+    const auto inv2_k13 = float((-k2[2]*k2[4] + k2[1]*k2[5]) / (k2[0]*k2[4]));
+    const auto inv2_k22 = float(1 / k2[4]);
+    const auto inv2_k23 = float(-k2[5] / k2[4]);
+
+    const int num_pts = points.rows;
+    norm_points = Mat (num_pts, 4, points.type());
+    auto * norm_points_ = (float *) norm_points.data;
+
+    for (int i = 0; i < num_pts; i++) {
+        const int idx = 4*i;
+        (*norm_points_++) = inv1_k11 * points_[idx  ] + inv1_k12 * points_[idx+1] + inv1_k13;
+        (*norm_points_++) =                             inv1_k22 * points_[idx+1] + inv1_k23;
+        (*norm_points_++) = inv2_k11 * points_[idx+2] + inv2_k12 * points_[idx+3] + inv2_k13;
+        (*norm_points_++) =                             inv2_k22 * points_[idx+3] + inv2_k23;
+    }
+}
+
+/*
+ * K is 3x3 intrinsic matrix
+ * points is matrix of size |N| x 5, first two columns are image points [u_i, v_i]
+ * calib_norm_pts are  K^-1 [u v 1]^T / ||K^-1 [u v 1]^T||
+ */
+void Utils::calibrateAndNormalizePointsPnP (const Mat &K, const Mat &pts, Mat &calib_norm_pts) {
+    const auto * const points = (float *) pts.data;
+    const auto * const k = (double *) K.data;
+    const auto inv_k11 = float(1 / k[0]);
+    const auto inv_k12 = float(-k[1] / (k[0]*k[4]));
+    const auto inv_k13 = float((-k[2]*k[4] + k[1]*k[5]) / (k[0]*k[4]));
+    const auto inv_k22 = float(1 / k[4]);
+    const auto inv_k23 = float(-k[5] / k[4]);
+
+    const int num_pts = pts.rows;
+    calib_norm_pts = Mat (num_pts, 3, pts.type());
+    auto * calib_norm_pts_ = (float *) calib_norm_pts.data;
+
+    for (int i = 0; i < num_pts; i++) {
+        const int idx = 5 * i;
+        const float k_inv_u = inv_k11 * points[idx] + inv_k12 * points[idx+1] + inv_k13;
+        const float k_inv_v =                         inv_k22 * points[idx+1] + inv_k23;
+        const float norm = 1.f / sqrtf(powf(k_inv_u, 2) + powf(k_inv_v, 2) + 1);
+        (*calib_norm_pts_++) = k_inv_u * norm;
+        (*calib_norm_pts_++) = k_inv_v * norm;
+        (*calib_norm_pts_++) =           norm;
+    }
+}
+
+/*
+ * decompose Projection Matrix to calibration, rotation and translation
+ * Assume K = [fx  0   tx
+ *             0   fy  ty
+ *             0   0   1]
+ */
+void Utils::decomposeProjection (const Mat &P, Mat &K_, Mat &R, Mat &t) {
```

Comment:
Wondering if [`decomposeProjectionMatrix()`](https://docs.opencv.org/4.4.0/d9/d0c/group__calib3d.html#gaaae5a7899faa1ffdf268cd9088940248) could be used instead?

Because if `decomposeProjection` is better, `decomposeProjectionMatrix()` could be updated maybe?

Reply:
the current OpenCV version of decomposeProjectionMatrix() assumes non-zero k_12. However, in most cases and solvePnP assume k_12 = 0. So, as I wrote before if we use P6P RANSAC (without prior knowledge of calibration) we can find projection matrix which will be decomposed into unknown intrinsic, rotation and translation. Also, decomposeProjectionMatrix() returns K, rotation and center C: for example, if I want to know translation I have to normalize C (which is 4D) and then multiply it by rotation, but my function already outputs translation.

Reply:
Thanks for the explanation.